### ISSUE TITLE
clickatell question

### ISSUE THREAD

**Issue by WeeseNXT** [nextcloud/twofactor_gateway#359]
So I have a clickatell account and have everything set up the way it needs to be but it ask me for my "central.clickatell.com" API, UN and Password, my issue is this is for accounts created prior to 2016. My account is at "portal.clickatell.com" can I make this app communicate with the new site instead? or do I have to have one of the old accounts?

**Comment by ChristophWurst** [comment]
@killerbees19 can you help with this? (ref https://github.com/nextcloud/twofactor_gateway/pull/333)

**Comment by killerbees19** [comment]
Sadly I don't have an account at the new site yet. My credits are only available at the old site.

But I think it should be possible to create a 2nd provider for the new site. According to the [docs](https://www.clickatell.com/developers/api-documentation/http-api-send-message/) the format differs only in one small part:

```ini
# Clickatell Central
https://api.clickatell.com/http/sendmsg?user=%s&password=%s&api_id=%u&to=%s&text=%s

# Clickatell Portal
https://platform.clickatell.com/messages/http/send?apiKey=%s&to=%s&content=%s
```

`apiKey` instead of `user`/`password`/`api_id`.

@WeeseNXT Can you confirm this information? And would you test a patch for the new site? Maybe I could implement it, but I'm unable to test it by myself at the moment.

**Comment by killerbees19** [comment]
@WeeseNXT Please test PR #360

Thanks!

**Comment by WeeseNXT** [comment]
That is correct, yes. I would certainly be willing to test it but I am a noob to all fo this, so forgive me if I need step by step instructions on the install or patch, etc. Thank you in advance for your reply. Im excited to test the new portal!

**Comment by WeeseNXT** [comment]
I'm on Nextcloud 19
Ubuntu server 18.04 LTS

Is there any relevant add-ons to nextcloud you would need to know about?

**Comment by killerbees19** [comment]
@WeeseNXT Do you know how to apply a patch/diff file? (_And revert it if needed…_)

**Comment by WeeseNXT** [comment]
I do not, I'm sorry, but looking to learn as much as I can.

**Comment by WeeseNXT** [comment]
I apologize I'm new to all of this, trying to learn.

**Comment by killerbees19** [comment]
@WeeseNXT Do you have SSH access to your account to execute custom commands in the Nextcloud directory?

**Comment by WeeseNXT** [comment]
Yes I do

**Comment by WeeseNXT** [comment]
It is my privately owned server at my house. I have 100% access to everything

**Comment by killerbees19** [comment]
Ok, then login to a shell with the right user, who's able to update Nextcloud files.

```bash
# Change to directory of this app (CHANGE PATH!)
cd /path/to/your/nextcloud/apps/twofactor_gateway/

# Create a backup (CHANGE PATH!)
tar cfa /where/to/store/backup.tar.gz ./

# Download patch...
wget -O /tmp/nc.patch https://github.com/froonix/twofactor_gateway/commit/b99fa6ea08be705cc03e0f38360bd77f2e4764b0.patch

# Test if patch could be applied...
patch --dry-run -p1 < /tmp/nc.patch

# No error? Apply patch...
patch -p1 < /tmp/nc.patch
```

To restore the patch later:
```bash
# Repeat the steps from above, but use this patch command instead:
patch -R -p1 < /tmp/nc.patch
```

Proceed with `occ twofactorauth:gateway:configure sms` as usual. You should see `clickatellportal` in the list.

**Comment by WeeseNXT** [comment]
@killerbees19 
This is what it looks like.

root@myservernamehere:~# sudo nextcloud.occ twofactorauth:gateway:configure sms                                                        Please choose a SMS provider (websms, playsms, clockworksms, puzzelsms, ecallsms, voipms, huawei_e3531, spryng, sms77io, ovh, clickatellcentral, clickatellportal

When I type clickatellportal it returns

'Invalid provider clickatell portal'

I'm no coding expert as previously stated so excuse me if I sound silly, lol, but I did notice that 'clicksend' is missing from the end of the list as well as the ')' so I feel like some part of the code broke and part of the program is missing. all of the other providers return normal functions. To test my theory I also tried 'clicksend' it returns 'Invalid provider' as well. 

- PRD

**Comment by WeeseNXT** [comment]
![PSX_20200705_213807](https://user-images.githubusercontent.com/67760285/86547715-e5d77380-bf07-11ea-85e9-10a5662189d8.jpg)

**Comment by WeeseNXT** [comment]
![sketch1593999755977](https://user-images.githubusercontent.com/67760285/86547918-83cb3e00-bf08-11ea-8bf6-7a3fcfdcb39a.png)

**Comment by killerbees19** [comment]
You're right, there's something wrong. I'll check that later and come back to you. Sorry! :-(

`clicksend` is missing because I've reformatted the patch to only include the changes between v16.0 release of the app and `clickatellportal` patch.

**Comment by killerbees19** [comment]
@WeeseNXT By bad, I forgot some lines while reformatting the patch…

**Please revert/remove the old patch as mentioned [before](https://github.com/nextcloud/twofactor_gateway/issues/359#issuecomment-653916359)!** :-)

Now try this one:
```bash
# Download patch...
wget -O /tmp/nc-2.patch https://github.com/froonix/twofactor_gateway/commit/9f09f7887bbfe5bed6d6270769ebc5988cde337c.patch

# Test if patch could be applied...
patch --dry-run -p1 < /tmp/nc-2.patch

# No error? Apply patch...
patch -p1 < /tmp/nc-2.patch
```

**Comment by WeeseNXT** [comment]
@killerbees19, this one prompts me for the API key, but when I enter it, it just ends the program and sends me back to a normal command line prompt, does not prompt me for my UN and password. I did test it just to make sure the portal didn't ONLY need the api, no luck.

![example](https://user-images.githubusercontent.com/67760285/86591563-d1bf6080-bf5f-11ea-9b5f-15a3930983a1.png)

**Comment by WeeseNXT** [comment]
I found this in the code library on my account and I'm not sure if this helps or not

Example showing how to send API call using CURL tool:

CURL CALL:
==========

curl "https://platform.clickatell.com/messages/http/send?apiKey=TLKC2f0xRriDorZdvcx05A==&to=14236183355&content=Test+message+text&from=1856-318-4169"


Example showing how to send API call from any Javascript environment:

`var xhr = new XMLHttpRequest();
xhr.open("GET", "https://platform.clickatell.com/messages/http/send?apiKey=TLKC2f0xRriDorZdvcx05A==&to=14236183355&content=Test+message+text&from=1856-318-4169", true);
xhr.onreadystatechange = function(){
    if (xhr.readyState == 4 && xhr.status == 200) {
        console.log('success');
    }
};
xhr.send();`

**Comment by killerbees19** [comment]
That's the expected behavior, it asks only for the API-Key. There's no username/password for the new API.

What's the output of `occ twofactorauth:gateway:status`? It should display something like:

```
Signal gateway: not configured
SMS gateway: configured
Telegram gateway: not configured
```

Does it work on the command line?

```bash
curl --remote-header-name 'https://platform.clickatell.com/messages/http/send?apiKey=YOURAPIKEY&to=TESTRECIPIENT&content=Hello'; echo
```

Replace `YOURAPIKEY` with your API-Key and `TESTRECIPIENT` with your own mobile phone number. What's the output of this command?

(**This command will cost you money!**)

**Comment by killerbees19** [comment]
btw: Does your own API-Key looks like `TLKC2f0xRriDorZdvcx05A==` too? With two equal signs at the end?

**Comment by killerbees19** [comment]
Ok, I've registered a new Clickatell account now and used the **Test integration** (Sandbox) feature. Works like a charm via Nextcloud. SMS code received at first try…

@WeeseNXT Are you sure you've created the right integration type? I think it's important to select **One-way messaging** at **Messaging type**.

![Bildschirmfoto_2020-07-06_16-00-02](https://user-images.githubusercontent.com/21372289/86601636-db69b800-bfa1-11ea-9430-8fd5c77372a6.png)

The `from` argument at your example looks like you've setup two-way messaging.

**Comment by WeeseNXT** [comment]
Sorry I've taken so long to reply. Work has been a mess today (on a 24 hr shift on the ambulance).
In the US two way is required to send pages via sms according to clickatell and other sites I'veread on. I've registered a return number on clickatell. It's throwing the error. "Invalid return number" I'm going to adjust some settings on my API integration and get back to you on what it says.

**Comment by WeeseNXT** [comment]
So I have tested the curl call in command line, as I said two way is required in the US when I do the curl call in cmd line with my "from number" it works perfectly and I can send whatever page I would like, content wise. However, when I attempt the two-factor auth in Nextcloud, no luck. Is there a way for me to look at an error report? Nextcloud just states that it's been sent.

**Comment by killerbees19** [comment]
> In the US two way is required to send pages via sms according to clickatell and other sites I've read on.

@WeeseNXT Is there a link/source for this statement? :-)

I mean: It's possible to change the code and ask for a source number too. And just include it in the request, if not empty. But I'd only change it if it's **really** required.

btw: It seems like currently there's no error check for some plugins, including clickatellcentral and clickatellportal. So a failed request is silently ignored. I'll fix that later.

**Comment by WeeseNXT** [comment]
The other websites I tried have said the same thing. Clockwork requires you email customer service to even get an integration for the US. Since my server is small time they weren't willing to do it for me since it isnt a high volume business. I can get you the emails and more links if you require, but here is a screenshot from my integration.

![20200707_185335](https://user-images.githubusercontent.com/67760285/86852964-a7ff5a00-c083-11ea-9511-2aba57ddd240.jpg)

**Comment by killerbees19** [comment]
Thanks for the screenshot! :-)

Ok, I'll look into this again later or tomorrow and prepare a new patch.

**Comment by killerbees19** [comment]
@WeeseNXT Sadly I can't test two-way messaging at Clickatell. But I think it should work now…

**Please revert/remove the last patch again!**

Now try this one:
```bash
# Download patch...
wget -O /tmp/nc-3.patch https://github.com/nextcloud/twofactor_gateway/compare/v0.16.0...froonix:test/issue/359/3.diff

# Test if patch could be applied...
patch --dry-run -p1 < /tmp/nc-3.patch

# No error? Apply patch...
patch -p1 < /tmp/nc-3.patch
```

`occ twofactorauth:gateway:configure sms` asks for a from number now:
```
Please choose a SMS provider (websms, playsms, clockworksms, puzzelsms, ecallsms, voipms, huawei_e3531, spryng, sms77io, ovh, clickatellcentral, clickatellportal): clickatellportal
Please enter your portal.clickatell.com API-Key: abcdef==
Please enter your sender number for two-way messaging. Leave it empty for one-way messaging: 436801234567890
```

Please experiment with the format of mobile numbers. I think they should be in international format, without leading zeros or plus sign. (_e.g. `436801234567890` for Austria [43] and 6801234567890 as the mobile number_)

I've implemented a simple error check too, but the detailed error message will be hidden.

@ChristophWurst Any plans to log SmsTransmissionException messages? Currently it's impossible to see the details of failed HTTP requests without throwing other exceptions and breaking the user experience. Or am I missing something?

**Comment by WeeseNXT** [comment]
I'm not sure what the deal is with this. Sorry, lol. I followed all of the steps and presumed it would work the exact same.

![sketch2134](https://user-images.githubusercontent.com/67760285/86870926-4a7d0480-c0a7-11ea-9834-bdbee317ce11.png)

**Comment by WeeseNXT** [comment]
That is where I remembered patching it to, to begin with. @killerbees19

**Comment by killerbees19** [comment]
@WeeseNXT Maybe you've rebootet in the meantime? Or there's a cleanup task for /tmp like tmpreaper.

However, just download the old patch again and then revert it:

```bash
wget -O /tmp/nc-2.patch https://github.com/froonix/twofactor_gateway/commit/9f09f7887bbfe5bed6d6270769ebc5988cde337c.patch
patch -R -p1 < /tmp/nc-2.patch
```

(Please double-check the commands, I'm writing this at my smartphone! ^^)

**Comment by WeeseNXT** [comment]
I did reboot, didn't realize that effected the patch. Learning more and more everyday. So I should just go from square one in that case?

**Comment by WeeseNXT** [comment]
@killerbees19, since I'm a command-line noob and I kept getting errors. What I did was uninstall the two-factor gateway and associated apps and reinstalled it I did your new patch following the instructions from scratch. It works like a charm. I just want to say I appreciate your willingness to teach and patch this program. Even though I couldn't offer any help as far as coding, etc. I hope that this will be applied to future editions and that this helped improve the app.

![Screenshot 2020-07-08 at 11 42 24](https://user-images.githubusercontent.com/67760285/86940603-058db800-c111-11ea-9e3b-c39b101d4753.png)

If you'd like me to test any future patches I would be willing to run some things on my server to test them, since it's small and mostly a hobby and extra storage for me. If I restart from here, it shouldn't have any effect on the current patch, right?

**Comment by killerbees19** [comment]
> I did reboot, didn't realize that effected the patch. 

/tmp will be flushed at shutdown/reboot/boot 😉

Sometimes it's even a tmpfs (stored in RAM).

Thanks for the feedback and testing! I'll finish the PR, hopefully it'll be integrated into next release of this app.

The patch should stay available at your Nextcloud installation as long there's no update for the 2FA app. And as long as it's not uninstalled. But I'd recommend to download it and save it somewhere. (Outside of /tmp 😃)

**Comment by vitormattos** [comment]
closed by #360
